File: app/src/main/java/com/andb/apps/trails/download/FileDownloader.java
package com.andb.apps.trails.download;

import android.os.AsyncTask;
import android.os.Environment;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;

public class FileDownloader {
    private static final int  MEGABYTE = 1024 * 1024;

    private static void downloadFromURL(String fileUrl, File directory){
        try {

            URL url = new URL(fileUrl);
            HttpURLConnection urlConnection = (HttpURLConnection)url.openConnection();
            urlConnection.connect();

            InputStream inputStream = urlConnection.getInputStream();
            FileOutputStream fileOutputStream = new FileOutputStream(directory);
            int totalSize = urlConnection.getContentLength();

            byte[] buffer = new byte[MEGABYTE];
            int bufferLength = 0;
            while((bufferLength = inputStream.read(buffer))>0 ){
                fileOutputStream.write(buffer, 0, bufferLength);
            }
            fileOutputStream.close();
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (MalformedURLException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static File downloadFile(String fileUrl, String fileName){
        String extStorageDirectory = Environment.getExternalStorageDirectory().toString();
        File folder = new File(extStorageDirectory, "SkiMaps");
        folder.mkdir();

        File pdfFile = new File(folder, fileName);

        try{
            if(pdfFile.exists()){
                return pdfFile;
            }else {
                pdfFile.createNewFile();
                FileDownloader.downloadFromURL(fileUrl, pdfFile);
            }
        }catch (IOException e){
            e.printStackTrace();
        }
        return pdfFile;
    }


}

File: app/src/main/java/com/andb/apps/trails/download/FileDownloader.kt
package com.andb.apps.trails.download

import android.app.DownloadManager
import android.content.Context
import android.content.Context.DOWNLOAD_SERVICE
import android.net.Uri
import android.util.Log
import com.andb.apps.trails.utils.filenameFromURL
import java.io.File
import java.io.FileNotFoundException
import java.io.FileOutputStream
import java.io.IOException
import java.net.HttpURLConnection
import java.net.MalformedURLException
import java.net.URL


object FileDownloader {
    private const val MEGABYTE = 1024 * 1024

    private fun downloadFromURL(fileUrl: String, directory: File) {
        try {

            val url = URL(fileUrl)
            val urlConnection = url.openConnection() as HttpURLConnection
            urlConnection.connect()

            val inputStream = urlConnection.inputStream
            val fileOutputStream = FileOutputStream(directory)

            val buffer = ByteArray(MEGABYTE)
            var bufferLength = inputStream.read(buffer)
            while (bufferLength > 0) {
                fileOutputStream.write(buffer, 0, bufferLength)
                bufferLength = inputStream.read(buffer)
            }
            fileOutputStream.close()
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        } catch (e: MalformedURLException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }

    }

    fun downloadFile(context: Context, fileUrl: String): File {
        val folder = File(context.filesDir.toString(), "PDFs")
        folder.mkdir()

        val pdfFile = File(folder, filenameFromURL(fileUrl))

        try {
            if (pdfFile.exists()) {
                return pdfFile
            } else {
                pdfFile.createNewFile()
                downloadFromURL(fileUrl, pdfFile)
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }

        return pdfFile
    }

    /*fun downloadFileExternal(context: Context, url: String, path: String){
        val pathToSave = context.getExternalFilesDir(null)
        pathToSave.mkdir()
        val fileToSave = File(pathToSave, path)

        if(fileToSave.exists()) return

        val possibleFile = File(File(context.filesDir.toString(), "PDFs"), filenameFromURL(url))
        if(possibleFile.exists()){
            val input = FileInputStream(possibleFile).channel
            val output = FileOutputStream(fileToSave).channel
            output.transferFrom(input, 0, input.size())
            input.close()
            output.close()
            return
        }

        downloadFromURL(url, fileToSave)
    }*/

    fun downloadFileExternal(context: Context, url: String, areaName: String, year: Int) {

        Log.d("downloadFileExternal", "start")

        val path = "${areaName}_${year}_${filenameFromURL(url)}"

        Log.d("downloadFileExternal", "path: $path")

        val pathToSave = context.getExternalFilesDir(null)
        val fileToSave = File(pathToSave, path)

        val notificationTitle = String.format(context.resources.getString(com.andb.apps.trails.R.string.downloading_notification_title), areaName, year)

        val request = DownloadManager.Request(Uri.parse(url))
            .setTitle(notificationTitle)
            .setDescription(path)
            .setDestinationUri(Uri.fromFile(fileToSave))
            .setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE_NOTIFY_COMPLETED)

        val downloadManager = context.getSystemService(DOWNLOAD_SERVICE) as DownloadManager
        val id = downloadManager.enqueue(request)

        Log.d("downloadFileExternal", "downloading id: $id")
    }
}


